File: server/server-core/src/main/java/com/simple/castle/server/tcp/ServerStarter.java
package com.simple.castle.server.tcp;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class ServerStarter {

    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final ServerListener serverListener;

    public ServerStarter(ServerListener serverListener) {
        this.serverListener = serverListener;
    }

    public void start() {
        executor.submit(serverListener);
    }

    public void stop() {
        System.out.println("Waiting for shutdown listener");
        try {
            serverListener.dispose();
            System.out.println("Disposed");

            if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
                executor.shutdownNow();
            }

            System.out.println("Done");
        } catch (InterruptedException e) {
            System.out.println("Interrupted");
        }
    }

}

File: server/server-core/src/main/java/com/simple/castle/server/kt/core/ServerStarter.kt
package com.simple.castle.server.kt.core

import com.badlogic.gdx.Gdx
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class ServerStarter(private val serverListener: ServerListener) {
    private val serverListenerService = Executors.newSingleThreadExecutor()
    fun start() {
        serverListenerService.submit(serverListener)
    }

    fun stop() {
        Gdx.app.log("ServerStarter", "Waiting for shutdown listener")
        try {
            serverListener.stop()
            serverListener.dispose()
            serverListenerService.shutdown()
            if (!serverListenerService.awaitTermination(5, TimeUnit.SECONDS)) {
                serverListenerService.shutdownNow()
            }
        } catch (e: InterruptedException) {
            Gdx.app.log("ServerStarter", e.message)
        }
        Gdx.app.log("ServerStarter", "Server was stopped")
    }

}